src/index/updater.rs

```diff
@@ -96,27 +98,46 @@ impl Updater {
       Some(progress_bar)
     };
 
-    let rx = Self::fetch_blocks_from(index, self.height, self.index_sats)?;
-
-    let (mut outpoint_sender, mut value_receiver) = Self::spawn_fetcher(index)?;
+    let rx = Self::fetch_blocks_from(self.index, self.height, self.index_sats)?;
 
     let mut uncommitted = 0;
-    let mut value_cache = HashMap::new();
     while let Ok(block) = rx.recv() {
-      self.index_block(
-        index,
-        &mut outpoint_sender,
-        &mut value_receiver,
-        &mut wtx,
-        block,
-        &mut value_cache,
-      )?;
+      if self.index.client.get_block_count()? - self.index.block_height()?.unwrap_or(Height(0)).n()
+        < 50
+      {
+        log::info!("not in initial sync; starting to create savepoints");
+        self.initial_sync = false;
+
+        if self.savepoints.len() >= 6 {
+          dbg!("here");
+          wtx.delete_persistent_savepoint(self.savepoints.pop_front().unwrap())?;
+        }
+
+        let savepoints: Vec<u64> = wtx.list_persistent_savepoints()?.into_iter().collect();
+        dbg!(&savepoints);
+        // self.savepoints.push_back(wtx.ephemeral_savepoint()?);
+
+        if let Some(prev_height) = self.height.checked_sub(1) {
+          let prev_hash = self.index.client.get_block_hash(prev_height)?;
+          dbg!(&prev_hash);
+          if prev_hash != block.header.prev_blockhash {
+            log::info!("reorg detected at or before {prev_height}; rolling back index");
+            self.index.reorged.store(true, atomic::Ordering::Relaxed);
+
+            let savepoint = wtx.get_persistent_savepoint(self.savepoints.pop_front().unwrap())?;
+            wtx.restore_savepoint(&savepoint)?;
+            self.savepoints.clear();
+          }
+        }
```

Comment:
This logic should probably live before the create savepoint logic, otherwise we create a savepoint and immediately rollback to an older one.

Reply:
your right, I put it into the Index now